swarm/src/registry.rs

```diff
@@ -0,0 +1,583 @@
+            xs.swap(j, j - 1)
+        }
+    }
+}
+
+impl NetworkBehaviour for Addresses {
+    type ConnectionHandler = crate::dummy::ConnectionHandler;
+
+    type ToSwarm = Void;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: crate::ConnectionId,
+        peer: libp2p_identity::PeerId,
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<crate::THandler<Self>, crate::ConnectionDenied> {
+        Ok(crate::dummy::ConnectionHandler)
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _connection_id: crate::ConnectionId,
+        peer: libp2p_identity::PeerId,
+        addr: &Multiaddr,
+        role_override: libp2p_core::Endpoint,
+    ) -> Result<crate::THandler<Self>, crate::ConnectionDenied> {
+        Ok(crate::dummy::ConnectionHandler)
+    }
+
+    fn on_swarm_event(&mut self, event: crate::FromSwarm<Self::ConnectionHandler>) {
+        match event {
+            crate::FromSwarm::ConnectionEstablished(_) => todo!(),
+            crate::FromSwarm::ConnectionClosed(_) => todo!(),
+            crate::FromSwarm::AddressChange(_) => todo!(),
+            crate::FromSwarm::DialFailure(_) => todo!(),
+            crate::FromSwarm::ListenFailure(_) => todo!(),
+            crate::FromSwarm::NewListener(_) => todo!(),
+            crate::FromSwarm::NewListenAddr(_) => todo!(),
+            crate::FromSwarm::ExpiredListenAddr(_) => todo!(),
+            crate::FromSwarm::ListenerError(_) => todo!(),
+            crate::FromSwarm::ListenerClosed(_) => todo!(),
+            crate::FromSwarm::NewExternalAddrCandidate(NewExternalAddrCandidate { addr }) => {
+                match self.add(addr.clone(), AddressScore::Finite(1)) {
+                    AddAddressResult::Inserted { expired } => {
+                        todo!("emit ToSwarm::NewExternalAddr(addr) and ToSwarm::ExternalAddrExpired(expired)")
+                    }
+                    AddAddressResult::Updated { expired } => {
+                        todo!("emit ToSwarm::ExternalAddrExpired(expired)")
+                    }
+                };
+            }
+            crate::FromSwarm::ExternalAddrConfirmed(ExternalAddrConfirmed { addr }) => {
+                self.add(addr.clone(), AddressScore::Infinite);
+            }
+            crate::FromSwarm::ExternalAddrExpired(ExternalAddrExpired { addr }) => {
+                self.remove(addr);
+            }
+        }
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        _peer_id: libp2p_identity::PeerId,
+        _connection_id: crate::ConnectionId,
+        event: crate::THandlerOutEvent<Self>,
+    ) {
+        void::unreachable(event)
+    }
+
+    fn poll(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+        _params: &mut impl crate::PollParameters,
+    ) -> std::task::Poll<crate::ToSwarm<Self::ToSwarm, crate::THandlerInEvent<Self>>> {
+        Poll::Pending
+    }
+}
```

Comment:
Note the only addition to the `v0.51` version, namely the `NetworkBehaviour` implementation that would allow one to simply plug this into ones `NetworkBehaviour` tree.